cachito/web/models.py

```diff
@@ -648,6 +650,31 @@ def __repr__(self):
             self.id, self.state_name, self.request_id
         )
 
+    @classmethod
+    def get_final_states_query(cls):
+        """Return query that filters complete/failed states and includes a 'duration' column."""
+        initial_state = aliased(cls)
+        return (
+            db.session.query(
+                cls.request_id,
+                cls.state,
+                cls.state_reason,
+                cls.updated,
+                func.extract(
+                    "epoch",
+                    cls.updated.cast(TIMESTAMP) - func.min(initial_state.updated).cast(TIMESTAMP),
+                ).label("duration"),
+            )
+            .join(initial_state, initial_state.request_id == cls.request_id)
+            .filter(initial_state.state == RequestStateMapping.in_progress.value)
```

Comment:
A request has many `in_progress` states. Does this always default to the earliest one?

Reply:
We need there only func.min(initial_state.updated) and then we group by final state. Any other data from initial_state will not appear in result.